azure-iot-device/azure/iot/device/iothub/abstract_clients.py

```diff
@@ -393,66 +399,66 @@ def create_from_sastoken(cls, sastoken, **kwargs):
         return cls(mqtt_pipeline, http_pipeline)
 
     @abc.abstractmethod
-    def shutdown(self):
+    def shutdown(self) -> None:
         pass
 
     @abc.abstractmethod
-    def connect(self):
+    def connect(self) -> None:
         pass
 
     @abc.abstractmethod
-    def disconnect(self):
+    def disconnect(self) -> None:
         pass
 
     @abc.abstractmethod
-    def update_sastoken(self, sastoken):
+    def update_sastoken(self, sastoken: str) -> None:
         pass
 
     @abc.abstractmethod
-    def send_message(self, message):
+    def send_message(self, message: Union[Message, str]) -> None:
         pass
 
     @abc.abstractmethod
-    def receive_method_request(self, method_name=None):
+    def receive_method_request(self, method_name: Optional[str] = None) -> None:
         pass
 
     @abc.abstractmethod
-    def send_method_response(self, method_request, payload, status):
+    def send_method_response(self, method_request: MethodRequest, payload: Dict[str, JSONSerializable], status: int) -> None:
         pass
 
     @abc.abstractmethod
-    def get_twin(self):
+    def get_twin(self) -> Twin:
         pass
 
     @abc.abstractmethod
-    def patch_twin_reported_properties(self, reported_properties_patch):
+    def patch_twin_reported_properties(self, reported_properties_patch: TwinPatch) -> None:
         pass
 
     @abc.abstractmethod
-    def receive_twin_desired_properties_patch(self):
+    def receive_twin_desired_properties_patch(self) -> TwinPatch:
         pass
 
     @property
-    def connected(self):
+    def connected(self) -> bool:
         """
         Read-only property to indicate if the transport is connected or not.
         """
         return self._mqtt_pipeline.connected
 
     @property
-    def on_connection_state_change(self):
+    def on_connection_state_change(self) -> Callable[[Any], Any]:
```

Comment:
I'm somewhat surprised this passes static analysis, although I suppose that's due to only annotating the surface.

The issue is that this is too broad. It says you can set any callable that takes any number of arguments and returns any type of value. This is not true, and following this annotation would produce errors.

This should probably be FunctionOrCoroutine[[None], None], however, I'm not quite sure if that is valid for all versions of Python - it used to not be due to a bug, but I believe that has been fixed since I reported it? If it hasn't been, I can live with Callable[[Any], Any], but it's problematic.

Here and everywhere obviously (although the specific type annotations would change with handler definition)